### ISSUE TITLE
Order of declarations affects shadowing error reporting

### ISSUE THREAD

**Issue by apblack** [gracelang/minigrace#258]
This program
```
method foo(secret) { secret }
method secret { ... }
```
should report a shadowing error when method `foo` uses `secret` as the name of its parameter.  If the order of the declarations is reversed, the error is correctly reported:
```
method secret { ... }
method foo(secret) { secret }
```

> Syntax error: 'secret' cannot be redeclared because it is already declared in an enclosing module scope on line 1. Use a different name.

This bug occurs because the declarations are processed in the order written.  What ought to happen is that the lhs of all of the declarations are processed before the rhs of any of them.  Not clear how to do this in a visitor!

**Comment by kjx** [comment]
`What ought to happen is that the lhs of all of the declarations are processed before the rhs of any of them. Not clear how to do this in a visitor!`

which is why I didn't do it in a visitor.  Evaluating anything that can hold declarations (methods, blocks, objects, etc)  must be done in two steps: first build the declarations into the context, second evaluate their values.  That's two mutually recursive methods, so presumably two mutually recursive visitors.

**Comment by apblack** [comment]
> That's two mutually recursive methods 

No, just two passes at compile time; first build the symbol tables, and then check for shadowing. I don't know why I didn't see this before; it's anotter you have her example of not understanding how to do something until after you have done it.

**Comment by apblack** [comment]
@Kim entered a duplicate of this as #260.  That issue contains a larger example.
